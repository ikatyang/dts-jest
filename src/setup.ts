import * as ts from 'typescript';
import { RawConfig, Result, Target } from './definitions';
import { Runtime } from './runtime';
import { create_snapshots } from './utils/create-snapshots';
import { default_to } from './utils/default-to';
import { get_tsconfig } from './utils/get-tsconfig';

export const setup = (
  filename: string,
  raw_config: RawConfig,
  targets: Target[],
) => {
  const lines = targets.map(target => target.line);

  const flag = default_to(
    raw_config.type_format,
    ts.TypeFormatFlags.NoTruncation,
  );
  const program = ts.createProgram([filename], get_tsconfig(raw_config));

  const snapshots = create_snapshots(program, filename, lines, flag);
  const results = targets.reduce<{ [line: number]: Result }>(
    (current_results, target) => ({
      ...current_results,
      [target.line]: {
        ...target,
        ...snapshots[target.line],
      },
    }),
    {},
  );

  return new Runtime(results);
};
