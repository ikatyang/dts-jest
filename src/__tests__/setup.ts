jest.unmock('../setup');

import * as fs from 'fs';
import * as path from 'path';
import {
  IRuntimeMethods,
  ISelfConfig,
  ITriggerDescriptions,
} from '../definitions';
import {setup} from '../setup';

type IRuntimeInformation = {
  [T in keyof IRuntimeMethods]: any;
};

interface IRuntimeInformations {
  [line: number]: IRuntimeInformation;
}

const get_informations_from_setup_fixture = (
    fixture_relative_dirname: string,
    config: ISelfConfig = {},
    ): IRuntimeInformations => {
  const dirname = path.resolve(__dirname, fixture_relative_dirname);
  const filename = path.join(dirname, 'test.ts');
  const descriptions: ITriggerDescriptions = JSON.parse(
    fs.readFileSync(path.join(dirname, 'descriptions.json'), 'utf8'),
  );
  const runtime_methods = setup({filename}, config, descriptions);

  const lines = Object.keys(descriptions).map(line_str => +line_str);
  return lines.reduce<IRuntimeInformations>(
    (current_informations, line) =>
      ({
        ...current_informations,
        [line]: {
          report: runtime_methods.report({line}),
          snapshot: runtime_methods.snapshot({line}),
          description: runtime_methods.description({line}),
        },
      }),
    {},
  );
};

it('should setup correctly while all triggers are attachable', () => {
  expect(get_informations_from_setup_fixture('../../fixtures/setup/general')).toMatchSnapshot();
});

it('should throw error while unattachable trigger(s) existed', () => {
  expect(() => get_informations_from_setup_fixture('../../fixtures/setup/unattachable')).toThrowErrorMatchingSnapshot();
});

it('should setup correctly with type-detail = false', () => {
  expect(get_informations_from_setup_fixture('../../fixtures/setup/type-detail', {
    type_detail: false,
  })).toMatchSnapshot();
});

it('should setup correctly with type-detail = true', () => {
  expect(get_informations_from_setup_fixture('../../fixtures/setup/type-detail', {
    type_detail: true,
  })).toMatchSnapshot();
});
