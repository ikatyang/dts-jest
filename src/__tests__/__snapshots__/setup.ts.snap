// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should setup correctly 1`] = `
"trigger(line:16) -> number[]
trigger(line:17) -> <T>(a: T) => I<T>
trigger(line:18) -> <T>(a: T) => X<T>
trigger(line:19) -> <T>(a: T) => { x: T; }
trigger(line:20) -> <T, U>(a: T, b: U) => T | U
trigger(line:21) -> { (): void; (): string; }
trigger(line:22) -> number
trigger(line:23) -> I<number[]>
trigger(line:24) -> X<number[]>
trigger(line:25) -> { x: number[]; }
trigger(line:26) -> 1 | \\"\\"
trigger(line:27) -> (<T>(a: T) => { x: T; }) | 1
trigger(line:28) -> void
trigger(line:29) -> Supplied parameters do not match any signature of call target.
trigger(line:30) -> Argument of type '{}' is not assignable to parameter of type 'I<I<I<string>>>'.
  Property 'x' is missing in type '{}'.
"
`;
